Add methods for converting lockfile back to string

chmeliik opened the pull request:
Add to_file, to_str as counterparts to from_file, from_str

Signed-off-by: Adam Cmiel <acmiel@redhat.com>

chmeliik committed:
Add methods for converting lockfile back to string

Add to_file, to_str as counterparts to from_file, from_str

Signed-off-by: Adam Cmiel <acmiel@redhat.com>

chmeliik commented:
CC @nirzari @ejegrova (can't add you as reviewers)

chmeliik commented:
Will be useful for CLOUDBLD-3314

nirzari commented on pyarn/lexer.py:
```diff
@@ -35,6 +35,11 @@ def t_NUMBER(t):
 
 
+# Unquoted string regex (see t_STRING)
+# Docstrings cannot use string interpolation, this value is used by other modules
```
Nitpick
```suggestion
# Docstrings can not use string interpolation, this value is used by other modules
```

chmeliik replied:
I double-checked some sources, e.g. https://english.stackexchange.com/questions/4510/why-is-cannot-spelled-as-one-word
TLDR both "cannot" and "can not" are acceptable, but "cannot" is more common

nirzari replied:
lol _The modern way of writing can not_ , perhaps I'm old

nirzari commented on pyarn/lexer.py:
```diff
@@ -35,6 +35,11 @@ def t_NUMBER(t):
 
+# Unquoted string regex (see t_STRING)
+# Docstrings cannot use string interpolation, this value is used by other modules
+UNQUOTED_STRING = r'[a-zA-Z/.-]([^\s\n,]*[^\s\n,:])?'
```
Just out of curiosity, doesn't black prefer double quotes over single?

chmeliik replied:
It does, but this project does not (yet?) use `black` and most of the strings I've seen here use single quotes

nirzari replied:
Yep. that's why question out of curiosity. Perhaps in the future we could just start using it which will fix it all.

athos-ribeiro replied:
I did not bring black in for the first commits intentionally. I thought pyarn was quite small and I did not see any need to enforce styles in the sense black does it. +1 to bring it in if you feel it would be worth it.

athos-ribeiro replied:
Why is this defined in the lexer and not in `lockfile.py` itself? ?

chmeliik replied:
> Why is this defined in the lexer and not in `lockfile.py` itself? ?

I wanted this to be defined as close to `t_STRING` as possible in the hope that if someone changes the relevant part of the `t_STRING` regex, they will remember to update the `UNQUOTED_STRING` regex accordingly (it is just copy-pasted from there)

athos-ribeiro replied:
Sounds good!

nirzari commented:
Semantically it makes sense to me. I'd wait for one more approval but LGTM :+1:

athos-ribeiro commented on pyarn/lockfile.py:
```diff
@@ -88,3 +91,50 @@ def from_str(cls, lockfile_str):
+
+    def _dump(self, outfile):
+        # Does not preserve any comments, but this one is required
+        outfile.write('# yarn lockfile v1\n')
```
Should this be a constant somewhere?

I realized later after processing this string that it is only used for lockfile v1.

chmeliik replied:
Should the constant also be used when determining the file version instead of the regex?

athos-ribeiro replied:
Sounds good, yes!

chmeliik replied:
ok, changed :-)

athos-ribeiro commented on tests/test_lockfile.py:
```diff
@@ -165,3 +166,59 @@ def test_package_with_comma():
+    assert content == EXPECTED_CONTENT
+
+
+def test_roundtrip(all_test_files, tmp_path):
```
I like the test name! :rofl:

athos-ribeiro commented on tests/test_lockfile.py:
```diff
@@ -165,3 +166,59 @@ def test_package_with_comma():
+def test_roundtrip(all_test_files, tmp_path):
+    for test_file in all_test_files:
+        out_file = tmp_path / os.path.basename(test_file)
+        # Check that load -> dump -> load produces expected results
```
nice!

athos-ribeiro reviewed (commented):
LGTM

Just a couple questions/nitpicks inline

chmeliik committed:
Change how lockfile version is determined

Instead of looking for a '# yarn lockfile v\d+' comment, look only for
the literal '# yarn lockfile v1'. Only v1 uses this comment.

Signed-off-by: Adam Cmiel <acmiel@redhat.com>
